// @tripetto/runner-fabric/dist/esm/components/checkboxes/index.mjs
import { jsx, Fragment } from "react/jsx-runtime";
import { CheckboxFabric } from "../checkbox/index.mjs";

const CheckboxesFabric = (props) =>
    jsx(Fragment, {
        children: props.checkboxes.map(
            (checkbox, index) =>
                (props.view === "preview" || checkbox.label) &&
                jsx(
                    CheckboxFabric,
                    {
                        styles: props.styles,
                        label: checkbox.label,
                        description: checkbox.description,
                        required: checkbox.required,
                        disabled: checkbox.disabled,
                        readOnly: checkbox.readOnly,
                        error: checkbox.error,
                        tabIndex: checkbox.tabIndex || props.tabIndex,
                        tabSubmit: index + 1 === props.checkboxes.length,
                        value: checkbox.value,
                        ariaDescribedBy: props.ariaDescribedBy,
                        onChange: checkbox.onChange,
                        onAutoFocus: (index === 0 && props.onAutoFocus) || undefined,
                        onFocus: props.onFocus,
                        onBlur: props.onBlur,
                        onSubmit: props.onSubmit,
                        onCancel: (index === 0 && props.onCancel) || undefined,
                    },
                    checkbox.id || index
                )
        ),
    });

export { CheckboxesFabric };
