// @tripetto/runner-autoscroll/angular/esm2020/lib/runner.module.mjs
import { NgModule } from "@angular/core";
import { TripettoAutoscrollComponent } from "./runner.component";
import * as i0 from "@angular/core";
export class TripettoAutoscrollModule {}
TripettoAutoscrollModule.ɵfac = i0.ɵɵngDeclareFactory({
    minVersion: "12.0.0",
    version: "15.2.1",
    ngImport: i0,
    type: TripettoAutoscrollModule,
    deps: [],
    target: i0.ɵɵFactoryTarget.NgModule,
});
TripettoAutoscrollModule.ɵmod = i0.ɵɵngDeclareNgModule({
    minVersion: "14.0.0",
    version: "15.2.1",
    ngImport: i0,
    type: TripettoAutoscrollModule,
    declarations: [TripettoAutoscrollComponent],
    exports: [TripettoAutoscrollComponent],
});
TripettoAutoscrollModule.ɵinj = i0.ɵɵngDeclareInjector({
    minVersion: "12.0.0",
    version: "15.2.1",
    ngImport: i0,
    type: TripettoAutoscrollModule,
});
i0.ɵɵngDeclareClassMetadata({
    minVersion: "12.0.0",
    version: "15.2.1",
    ngImport: i0,
    type: TripettoAutoscrollModule,
    decorators: [
        {
            type: NgModule,
            args: [
                {
                    declarations: [TripettoAutoscrollComponent],
                    imports: [],
                    exports: [TripettoAutoscrollComponent],
                },
            ],
        },
    ],
});
