// @tripetto/runner-fabric/dist/es5/components/checkboxes/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckboxesFabric = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var checkbox_1 = require("../checkbox");
var CheckboxesFabric = function (props) {
    return (0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, {
        children: props.checkboxes.map(function (checkbox, index) {
            return (
                (props.view === "preview" || checkbox.label) &&
                (0, jsx_runtime_1.jsx)(
                    checkbox_1.CheckboxFabric,
                    {
                        styles: props.styles,
                        label: checkbox.label,
                        description: checkbox.description,
                        required: checkbox.required,
                        disabled: checkbox.disabled,
                        readOnly: checkbox.readOnly,
                        error: checkbox.error,
                        tabIndex: checkbox.tabIndex || props.tabIndex,
                        tabSubmit: index + 1 === props.checkboxes.length,
                        value: checkbox.value,
                        ariaDescribedBy: props.ariaDescribedBy,
                        onChange: checkbox.onChange,
                        onAutoFocus: (index === 0 && props.onAutoFocus) || undefined,
                        onFocus: props.onFocus,
                        onBlur: props.onBlur,
                        onSubmit: props.onSubmit,
                        onCancel: (index === 0 && props.onCancel) || undefined,
                    },
                    checkbox.id || index
                )
            );
        }),
    });
};
exports.CheckboxesFabric = CheckboxesFabric;
